src/NServiceBus.Core/Recoverability/RecoverabilityBehavior.cs

```diff
@@ -0,0 +1,93 @@
+﻿namespace NServiceBus
+{
+    using System;
+    using System.Threading.Tasks;
+    using Pipeline;
+
+    class RecoverabilityBehavior : Behavior<ITransportReceiveContext>
+    {
+        public RecoverabilityBehavior(FirstLevelRetriesHandler flrHandler, SecondLevelRetriesHandler slrHandler, MoveFaultsToErrorQueueHandler errorHandler,
+            bool flrEnabled, bool slrEnabled, bool runningWithTransactions)
+        {
+            this.flrHandler = flrHandler;
+            this.slrHandler = slrHandler;
+            this.errorHandler = errorHandler;
+
+            this.flrEnabled = flrEnabled;
+            this.slrEnabled = slrEnabled;
+            this.runningWithTransactions = runningWithTransactions;
+        }
+
+        public override async Task Invoke(ITransportReceiveContext context, Func<Task> next)
+        {
+            if (await errorHandler.HandleIfPreviouslyFailed(context).ConfigureAwait(false))
+            {
+                return;
+            }
+
+            try
+            {
+                await ExecuteWithRetries(context, next).ConfigureAwait(true);
+            }
+            catch (Exception ex)
+            {
+
+                if (runningWithTransactions)
+                {
+                    errorHandler.MarkForFutureHandling(context, ex);
+                }
+                else
+                {
+                    await errorHandler.MoveMessageToErrorQueue(context, ex).ConfigureAwait(false);
+                }
+            }
+        }
+
+        async Task ExecuteWithRetries(ITransportReceiveContext context, Func<Task> next)
+        {
+            if (slrEnabled && await slrHandler.HandleIfPreviouslyFailed(context).ConfigureAwait(false))
+            {
+                return;
+            }
+
+            try
+            {
+                await next().ConfigureAwait(false);
+            }
+            catch (MessageDeserializationException)
+            {
+                if (slrEnabled) //TODO: why do we have this thing here :/?
```

Comment:
not sure, but I think that was probably done in case the message was immediately sent to the error queue without a rollback to not contain any retries header. Doesn't seem to make much sense, especially since we're removing the headers anyway on the error queue action executor. I think we can remove this?